File: src-tauri/src/remote_server/viewer_page.js
(() => {
  "use strict";

  const statusElement = document.getElementById("status");
  const pathElement = document.getElementById("path");
  const textElement = document.getElementById("text");
  const imageElement = document.getElementById("image");
  const previewElement = document.getElementById("preview");
  const binaryElement = document.getElementById("binary");
  let acceptedSession = false;

  function setStatus(message, isError = false) {
    statusElement.textContent = message;
    statusElement.classList.toggle("error", isError);
    statusElement.hidden = false;
  }

  function hideContent() {
    textElement.hidden = true;
    imageElement.hidden = true;
    previewElement.hidden = true;
    binaryElement.hidden = true;
    imageElement.removeAttribute("src");
    previewElement.removeAttribute("srcdoc");
  }

  function formatBytes(value) {
    if (!Number.isFinite(value) || value < 0) return "Unknown size";
    if (value < 1024) return `${value} B`;
    const units = ["KiB", "MiB", "GiB"];
    let size = value / 1024;
    let unit = units[0];
    for (let index = 1; index < units.length && size >= 1024; index += 1) {
      size /= 1024;
      unit = units[index];
    }
    return `${size.toFixed(size >= 10 ? 1 : 2)} ${unit}`;
  }

  function render(payload) {
    hideContent();
    statusElement.hidden = true;
    pathElement.textContent = payload.path || "";
    pathElement.title = payload.path || "";
    document.title = payload.path
      ? `${payload.path} — Laymux`
      : "Laymux File Viewer";

    if (payload.kind === "text" && payload.previewDocument) {
      previewElement.setAttribute("sandbox", "");
      previewElement.srcdoc = payload.previewDocument;
      previewElement.hidden = false;
      return;
    }
    if (payload.kind === "text") {
      textElement.textContent = payload.content || "";
      textElement.hidden = false;
      if (payload.truncated)
        setStatus("Preview truncated at the Remote viewer limit.");
      return;
    }
    if (
      payload.kind === "image" &&
      /^data:image\//i.test(payload.dataUrl || "")
    ) {
      imageElement.src = payload.dataUrl;
      imageElement.hidden = false;
      return;
    }
    if (payload.kind === "binary") {
      binaryElement.textContent = `Binary or unsupported file · ${formatBytes(payload.size)}`;
      binaryElement.hidden = false;
      return;
    }
    throw new Error("Unsupported viewer response");
  }

  async function loadSession(session) {
    const body = { source: session.source };
    if (session.source === "path") body.path = session.path;
    const response = await fetch("/remote/v1/file-viewer/render", {
      method: "POST",
      headers: {
        authorization: `Bearer ${session.token}`,
        "content-type": "application/json",
        "x-laymux-remote-lease": session.leaseId,
      },
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      const body = await response.json().catch(() => ({}));
      throw new Error(
        body.error || `${response.status} ${response.statusText}`,
      );
    }
    render(await response.json());
  }

  window.addEventListener("message", (event) => {
    if (
      acceptedSession ||
      event.origin !== window.location.origin ||
      event.source !== window.opener
    ) {
      return;
    }
    const session = event.data;
    if (
      !session ||
      session.type !== "laymux:file-viewer-session" ||
      typeof session.token !== "string" ||
      !session.token ||
      typeof session.leaseId !== "string" ||
      !session.leaseId ||
      (session.source !== "current" && session.source !== "path") ||
      (session.source === "path" &&
        (typeof session.path !== "string" || !session.path.trim()))
    ) {
      return;
    }

    acceptedSession = true;
    window.opener = null;
    setStatus("Loading file…");
    loadSession(session).catch((error) => {
      hideContent();
      setStatus(error instanceof Error ? error.message : String(error), true);
    });
  });

  if (!window.opener) {
    setStatus("Open this viewer from a connected Laymux Remote page.", true);
    return;
  }
  window.opener.postMessage(
    { type: "laymux:file-viewer-ready" },
    window.location.origin,
  );
})();

File: src-tauri/src/remote_server/viewer_page.rs
use std::net::SocketAddr;

use axum::extract::{ConnectInfo, Request, State};
use axum::http::{header, HeaderValue};
use axum::response::{Html, IntoResponse, Response};

use crate::automation_server::ServerState;

use super::page::remote_page_gate;

const VIEWER_CSP: &str = "default-src 'none'; script-src 'self'; style-src 'unsafe-inline'; img-src data:; connect-src 'self'; base-uri 'none'; form-action 'none'; frame-src 'self'; frame-ancestors 'none'";

pub(super) async fn remote_viewer_page(
    State(server): State<ServerState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
    req: Request,
) -> Response {
    if let Some(response) = remote_page_gate(&server, addr, &req) {
        return response;
    }
    secure_viewer_response(Html(REMOTE_VIEWER_HTML).into_response())
}

pub(super) async fn remote_viewer_javascript(
    State(server): State<ServerState>,
    ConnectInfo(addr): ConnectInfo<SocketAddr>,
    req: Request,
) -> Response {
    if let Some(response) = remote_page_gate(&server, addr, &req) {
        return response;
    }
    let mut response = REMOTE_VIEWER_JS.into_response();
    response.headers_mut().insert(
        header::CONTENT_TYPE,
        HeaderValue::from_static("text/javascript; charset=utf-8"),
    );
    secure_viewer_response(response)
}

fn secure_viewer_response(mut response: Response) -> Response {
    let headers = response.headers_mut();
    headers.insert(header::CACHE_CONTROL, HeaderValue::from_static("no-store"));
    headers.insert(
        header::CONTENT_SECURITY_POLICY,
        HeaderValue::from_static(VIEWER_CSP),
    );
    headers.insert(
        header::REFERRER_POLICY,
        HeaderValue::from_static("no-referrer"),
    );
    headers.insert(
        header::X_CONTENT_TYPE_OPTIONS,
        HeaderValue::from_static("nosniff"),
    );
    response
}

const REMOTE_VIEWER_HTML: &str = include_str!("viewer_page.html");
const REMOTE_VIEWER_JS: &str = include_str!("viewer_page.js");

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn viewer_bootstrap_has_no_embedded_credentials_or_inline_script() {
        assert!(REMOTE_VIEWER_HTML.contains("Laymux File Viewer"));
        assert!(REMOTE_VIEWER_HTML.contains("/remote/viewer/viewer.js"));
        assert!(!REMOTE_VIEWER_HTML.contains("<script>"));
        assert!(!REMOTE_VIEWER_HTML.contains("token="));
        assert!(!REMOTE_VIEWER_HTML.contains("leaseId"));
    }

    #[test]
    fn viewer_script_accepts_one_exact_origin_opener_session() {
        assert!(REMOTE_VIEWER_JS.contains("laymux:file-viewer-ready"));
        assert!(REMOTE_VIEWER_JS.contains("laymux:file-viewer-session"));
        assert!(REMOTE_VIEWER_JS.contains("event.origin !== window.location.origin"));
        assert!(REMOTE_VIEWER_JS.contains("event.source !== window.opener"));
        assert!(REMOTE_VIEWER_JS.contains("window.opener = null"));
        assert!(REMOTE_VIEWER_JS.contains("x-laymux-remote-lease"));
        assert!(REMOTE_VIEWER_JS.contains("x-laymux-remote-file-viewer"));
        assert!(REMOTE_VIEWER_JS.contains("sandbox"));
    }

    #[test]
    fn viewer_script_never_copies_a_host_path_into_the_document_title() {
        assert!(!REMOTE_VIEWER_JS.contains("document.title ="));
    }

    #[test]
    fn sandboxed_preview_reports_when_its_source_was_truncated() {
        let branch_start = REMOTE_VIEWER_JS
            .find("if (payload.kind === \"text\" && payload.previewDocument)")
            .expect("preview branch");
        let branch_return = REMOTE_VIEWER_JS[branch_start..]
            .find("return;")
            .map(|offset| branch_start + offset)
            .expect("preview branch return");
        let branch = &REMOTE_VIEWER_JS[branch_start..branch_return];

        assert!(branch.contains("payload.truncated"));
        assert!(branch.contains("Preview truncated at the Remote viewer limit."));
    }

    #[test]
    fn viewer_csp_blocks_inline_and_remote_scripts() {
        assert!(VIEWER_CSP.contains("script-src 'self'"));
        assert!(!VIEWER_CSP.contains("script-src 'unsafe-inline'"));
        assert!(VIEWER_CSP.contains("frame-ancestors 'none'"));
    }
}
